File: leetcode/add-two-numbers.java
// https://leetcode.com/problems/add-two-numbers/

class Solution {

	// Provided ListNode class
	public class ListNode {
		int val;
		ListNode next;
		ListNode() {}
		ListNode(int val) { this.val = val; }
		ListNode(int val, ListNode next) { this.val = val; this.next = next; }
	}

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode combined = new ListNode();
        ListNode curr = combined;
        int carry = 0;
        
        while(l1 != null || l2 != null){
            int sum = (l1 != null ? l1.val : 0) + (l2 != null ? l2.val : 0) + carry;
            if(sum > 9){
                sum = sum%10;
                carry = 1;
            }else{
                carry = 0;
            }
            
            curr.val = sum;
            if(l1 != null) l1 = l1.next;
            if(l2 != null) l2 = l2.next;
            if(l1 != null || l2 != null){
                ListNode next = new ListNode();
                curr.next = next;
                curr = curr.next;
            }
        }
        
        if(carry == 1){
            curr.next = new ListNode(1);
        }
        
        return combined;
    }
}

File: leetcode/add_two_numbers.py
# https://leetcode.com/problems/add-two-numbers

from typing import Optional


# Provided ListNode class
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        one: str = ""
        while l1:
            one += str(l1.val)
            l1 = l1.next

        two: str = ""
        while l2:
            two += str(l2.val)
            l2 = l2.next

        val: str = str(int("".join(reversed(one))) + int("".join(reversed(two))))
        output: ListNode = ListNode()
        curr: ListNode = output

        for i in range(len(val) - 1, -1, -1):
            curr.val = val[i]
            if i > 0:
                temp: ListNode = ListNode()
                curr.next = temp
                curr = temp

        return output
